File: days/day_10/solutions/solution.ts
import * as fs from 'fs';

interface ParseResult {
    targetPattern: boolean[];
    buttons: number[][];
    joltages: number[];
}

function parseLine(line: string): ParseResult {
    const result: ParseResult = { targetPattern: [], buttons: [], joltages: [] };
    
    // Extract pattern: [.##.]
    const patternMatch = line.match(/\[([.#]+)\]/);
    if (!patternMatch) return result;
    const patternStr = patternMatch[1];
    result.targetPattern = patternStr.split('').map(c => c === '#');
    
    // Extract buttons: (1,3) (2) etc.
    const buttonMatches = line.matchAll(/\(([^)]*)\)/g);
    for (const match of buttonMatches) {
        const btnStr = match[1].trim();
        if (!btnStr) {
            result.buttons.push([]);
        } else {
            const lights = btnStr.split(',').map(x => parseInt(x.trim(), 10));
            result.buttons.push(lights);
        }
    }
    
    // Extract joltages: {3,5,4,7}
    const joltageMatch = line.match(/\{([^}]+)\}/);
    if (joltageMatch) {
        const joltageStr = joltageMatch[1];
        result.joltages = joltageStr.split(',').map(x => parseInt(x.trim(), 10));
    }
    
    return result;
}

function gaussianEliminationGF2(matrix: boolean[][], target: boolean[]): number | null {
    const numButtons = matrix.length;
    const numLights = target.length;
    
    // Create augmented matrix [A | b]
    const aug: boolean[][] = [];
    for (let i = 0; i < numLights; i++) {
        const row: boolean[] = [];
        for (let j = 0; j < numButtons; j++) {
            row.push(matrix[j] ? (matrix[j][i] || false) : false);
        }
        row.push(target[i]);
        aug.push(row);
    }
    
    // Gaussian elimination mod 2
    let pivotRow = 0;
    let pivotCol = 0;
    
    while (pivotRow < numLights && pivotCol < numButtons) {
        // Find pivot
        let pivotIdx: number | null = null;
        for (let i = pivotRow; i < numLights; i++) {
            if (aug[i][pivotCol]) {
                pivotIdx = i;
                break;
            }
        }
        
        if (pivotIdx === null) {
            pivotCol++;
            continue;
        }
        
        // Swap rows
        if (pivotIdx !== pivotRow) {
            [aug[pivotRow], aug[pivotIdx]] = [aug[pivotIdx], aug[pivotRow]];
        }
        
        // Eliminate
        for (let i = pivotRow + 1; i < numLights; i++) {
            if (aug[i][pivotCol]) {
                for (let j = 0; j < numButtons + 1; j++) {
                    aug[i][j] = aug[i][j] !== aug[pivotRow][j];
                }
            }
        }
        
        pivotRow++;
        pivotCol++;
    }
    
    // Check for inconsistency (0 = 1)
    for (let i = pivotRow; i < numLights; i++) {
        if (aug[i][numButtons] && !aug[i].slice(0, numButtons).some(x => x)) {
            return null; // No solution
        }
    }
    
    // Back substitution to find solution (set free variables to 0 to minimize)
    const solution: boolean[] = new Array(numButtons).fill(false);
    const usedRows = new Set<number>();
    
    // Process from bottom up
    for (let i = numLights - 1; i >= 0; i--) {
        // Find first non-zero column
        let pivotColIdx: number | null = null;
        for (let j = 0; j < numButtons; j++) {
            if (aug[i][j] && !usedRows.has(j)) {
                pivotColIdx = j;
                usedRows.add(j);
                break;
            }
        }
        
        if (pivotColIdx !== null) {
            // Calculate value
            let val = aug[i][numButtons];
            for (let j = pivotColIdx + 1; j < numButtons; j++) {
                if (aug[i][j] && solution[j]) {
                    val = !val;
                }
            }
            solution[pivotColIdx] = val;
        }
    }
    
    // Count number of presses
    return solution.filter(x => x).length;
}

function solvePart2ILP(buttons: number[][], joltages: number[], targetPattern: boolean[]): number | null {
    const numButtons = buttons.length;
    const numLights = joltages.length;
    
    // Base case: all joltages are 0
    if (joltages.every(j => j === 0)) {
        return 0;
    }
    
    // Bifurcate approach: try all parity solutions (buttons pressed 0 or 1 time)
    // that achieve the required parity, then divide by 2 and recurse
    const buttonMatrix: boolean[][] = Array(numButtons).fill(null).map(() => Array(numLights).fill(false));
    for (let i = 0; i < numButtons; i++) {
        for (const light of buttons[i]) {
            if (light >= 0 && light < numLights) {
                buttonMatrix[i][light] = true;
            }
        }
    }
    
    // Required parity: pattern ON = odd, pattern OFF = even
    const requiredParity = [...targetPattern];
    
    // Find minimum parity presses needed (Part 1 style)
    const parityToggles = [...requiredParity];
    const minParityPresses = gaussianEliminationGF2(buttonMatrix, parityToggles);
    
    if (minParityPresses !== null) {
        // Helper to generate combinations
        function* combinations(n: number, k: number): Generator<number[]> {
            if (k === 0) {
                yield [];
                return;
            }
            if (k > n) return;
            if (k === n) {
                yield Array.from({length: n}, (_, i) => i);
                return;
            }
            // Include first element
            for (const combo of combinations(n - 1, k - 1)) {
                yield [0, ...combo.map(x => x + 1)];
            }
            // Exclude first element
            for (const combo of combinations(n - 1, k)) {
                yield combo.map(x => x + 1);
            }
        }
        
        let bestResult: number | null = null;
        const maxParityToTry = numButtons;
        
        for (let k = minParityPresses; k <= maxParityToTry; k++) {
            // Pruning
            if (bestResult !== null && k > bestResult) {
                break;
            }
            
            for (const buttonCombo of combinations(numButtons, k)) {
                // Simulate pressing these buttons once
                const resultingJoltages = [...joltages];
                for (const btnIdx of buttonCombo) {
                    for (const light of buttons[btnIdx]) {
                        if (light >= 0 && light < numLights) {
                            resultingJoltages[light]++;
                        }
                    }
                }
                
                // Check if parity matches
                const resultingParity = resultingJoltages.map(j => j % 2 === 1);
                if (JSON.stringify(resultingParity) === JSON.stringify(requiredParity)) {
                    // Check if all remaining are even (can divide by 2)
                    if (resultingJoltages.every(j => j % 2 === 0)) {
                        // Divide by 2 and recurse
                        const halvedJoltages = resultingJoltages.map(j => Math.floor(j / 2));
                        const remainingResult = solvePart2ILP(buttons, halvedJoltages, targetPattern);
                        if (remainingResult !== null) {
                            const total = k + remainingResult * 2;
                            if (bestResult === null || total < bestResult) {
                                bestResult = total;
                            }
                        }
                    }
                }
            }
        }
        
        if (bestResult !== null) {
            return bestResult;
        }
    }
    
    // Fallback: divide-by-2 optimization
    if (joltages.every(j => j % 2 === 0) && joltages.some(j => j > 0)) {
        const halvedJoltages = joltages.map(j => Math.floor(j / 2));
        const result = solvePart2ILP(buttons, halvedJoltages, targetPattern);
        if (result !== null) {
            return result * 2;
        }
        return null;
    }
    
    // Final fallback: bounded DFS
    const maxJoltage = joltages.length > 0 ? Math.max(...joltages) : 0;
    
    function dfs(buttonIdx: number, currentJoltages: number[], pressesSoFar: number, best: number | null): number | null {
        if (buttonIdx >= numButtons) {
            if (currentJoltages.every((val, i) => val === joltages[i])) {
                return best === null ? pressesSoFar : Math.min(best, pressesSoFar);
            }
            return best;
        }
        
        if (best !== null && pressesSoFar >= best) {
            return best;
        }
        
        let bestResult = best;
        for (let presses = 0; presses <= maxJoltage; presses++) {
            if (best !== null && pressesSoFar + presses >= best) {
                break;
            }
            
            const newJoltages = [...currentJoltages];
            for (const light of buttons[buttonIdx]) {
                if (light >= 0 && light < numLights) {
                    newJoltages[light] += presses;
                }
            }
            
            if (newJoltages.some((val, i) => val > joltages[i])) {
                continue;
            }
            
            const result = dfs(buttonIdx + 1, newJoltages, pressesSoFar + presses, bestResult);
            if (result !== null) {
                bestResult = result;
            }
        }
        
        return bestResult;
    }
    
    const initialJoltages = Array(numLights).fill(0);
    return dfs(0, initialJoltages, 0, null);
}

function solve(inputData: string): [string, string] {
    const lines = inputData.trim().split('\n');
    if (lines.length === 0) {
        return ["0", "0"];
    }
    
    let part1Total = 0;
    let part2Total = 0;
    
    for (const line of lines) {
        if (!line.trim()) continue;
        
        const { targetPattern, buttons, joltages } = parseLine(line);
        if (targetPattern.length === 0) continue;
        
        const numLights = targetPattern.length;
        
        // Part 1: GF(2) linear system
        // Build incidence matrix: matrix[i][j] = true if button i toggles light j
        const buttonMatrix: boolean[][] = Array(buttons.length).fill(null).map(() => Array(numLights).fill(false));
        for (let i = 0; i < buttons.length; i++) {
            for (const light of buttons[i]) {
                if (light >= 0 && light < numLights) {
                    buttonMatrix[i][light] = true;
                }
            }
        }
        
        // Target: all start OFF, need to toggle to match pattern
        // Required toggles: target_pattern (1 = needs toggle, 0 = no toggle)
        const requiredToggles = [...targetPattern];
        
        const result = gaussianEliminationGF2(buttonMatrix, requiredToggles);
        if (result !== null) {
            part1Total += result;
        }
        
        // Part 2: Integer Linear Programming
        if (joltages.length === numLights) {
            const result2 = solvePart2ILP(buttons, joltages, targetPattern);
            if (result2 !== null) {
                part2Total += result2;
            }
        }
    }
    
    return [part1Total.toString(), part2Total.toString()];
}

const content = fs.readFileSync('../data/input.txt', 'utf-8');
const [part1, part2] = solve(content);
console.log(`Part 1: ${part1}`);
console.log(`Part 2: ${part2}`);

File: days/day_10/solutions/solution.rs
use std::fs;
use regex::Regex;

struct ParseResult {
    target_pattern: Vec<bool>,
    buttons: Vec<Vec<i32>>,
    joltages: Vec<i32>,
}

fn parse_line(line: &str) -> ParseResult {
    let mut result = ParseResult {
        target_pattern: Vec::new(),
        buttons: Vec::new(),
        joltages: Vec::new(),
    };
    
    // Extract pattern: [.##.]
    let pattern_re = Regex::new(r"\[([.#]+)\]").unwrap();
    if let Some(caps) = pattern_re.captures(line) {
        let pattern_str = caps.get(1).unwrap().as_str();
        result.target_pattern = pattern_str.chars().map(|c| c == '#').collect();
    }
    
    // Extract buttons: (1,3) (2) etc.
    let button_re = Regex::new(r"\(([^)]*)\)").unwrap();
    for caps in button_re.captures_iter(line) {
        let btn_str = caps.get(1).unwrap().as_str().trim();
        if btn_str.is_empty() {
            result.buttons.push(Vec::new());
        } else {
            let lights: Vec<i32> = btn_str.split(',')
                .map(|s| s.trim().parse().unwrap_or(0))
                .collect();
            result.buttons.push(lights);
        }
    }
    
    // Extract joltages: {3,5,4,7}
    let joltage_re = Regex::new(r"\{([^}]+)\}").unwrap();
    if let Some(caps) = joltage_re.captures(line) {
        let joltage_str = caps.get(1).unwrap().as_str();
        result.joltages = joltage_str.split(',')
            .map(|s| s.trim().parse().unwrap_or(0))
            .collect();
    }
    
    result
}

fn gaussian_elimination_gf2(matrix: &[Vec<bool>], target: &[bool]) -> Option<i32> {
    let num_buttons = matrix.len();
    let num_lights = target.len();
    
    // Create augmented matrix [A | b]
    let mut aug: Vec<Vec<bool>> = Vec::new();
    for i in 0..num_lights {
        let mut row = Vec::new();
        for j in 0..num_buttons {
            if j < matrix.len() && i < matrix[j].len() {
                row.push(matrix[j][i]);
            } else {
                row.push(false);
            }
        }
        row.push(target[i]);
        aug.push(row);
    }
    
    // Gaussian elimination mod 2
    let mut pivot_row = 0;
    let mut pivot_col = 0;
    
    while pivot_row < num_lights && pivot_col < num_buttons {
        // Find pivot
        let mut pivot_idx: Option<usize> = None;
        for i in pivot_row..num_lights {
            if aug[i][pivot_col] {
                pivot_idx = Some(i);
                break;
            }
        }
        
        if let Some(idx) = pivot_idx {
            // Swap rows
            if idx != pivot_row {
                aug.swap(pivot_row, idx);
            }
            
            // Eliminate
            for i in (pivot_row + 1)..num_lights {
                if aug[i][pivot_col] {
                    for j in 0..=num_buttons {
                        aug[i][j] ^= aug[pivot_row][j];
                    }
                }
            }
            
            pivot_row += 1;
        }
        pivot_col += 1;
    }
    
    // Check for inconsistency (0 = 1)
    for i in pivot_row..num_lights {
        if aug[i][num_buttons] && !aug[i][..num_buttons].iter().any(|&x| x) {
            return None; // No solution
        }
    }
    
    // Back substitution to find solution (set free variables to 0 to minimize)
    let mut solution = vec![false; num_buttons];
    let mut used_rows = std::collections::HashSet::new();
    
    // Process from bottom up
    for i in (0..num_lights).rev() {
        // Find first non-zero column
        let mut pivot_col_idx: Option<usize> = None;
        for j in 0..num_buttons {
            if aug[i][j] && !used_rows.contains(&j) {
                pivot_col_idx = Some(j);
                used_rows.insert(j);
                break;
            }
        }
        
        if let Some(idx) = pivot_col_idx {
            // Calculate value
            let mut val = aug[i][num_buttons];
            for j in (idx + 1)..num_buttons {
                if aug[i][j] && solution[j] {
                    val ^= true;
                }
            }
            solution[idx] = val;
        }
    }
    
    // Count number of presses
    Some(solution.iter().filter(|&&x| x).count() as i32)
}

fn solve_part2_ilp(buttons: &[Vec<i32>], joltages: &[i32], target_pattern: &[bool]) -> Option<i32> {
    let num_buttons = buttons.len();
    let num_lights = joltages.len();
    
    // Base case: all joltages are 0
    if joltages.iter().all(|&j| j == 0) {
        return Some(0);
    }
    
    // Bifurcate approach: try all parity solutions (buttons pressed 0 or 1 time)
    // that achieve the required parity, then divide by 2 and recurse
    let mut button_matrix = vec![vec![false; num_lights]; num_buttons];
    for (i, btn) in buttons.iter().enumerate() {
        for &light in btn {
            if light >= 0 && (light as usize) < num_lights {
                button_matrix[i][light as usize] = true;
            }
        }
    }
    
    // Required parity: pattern ON = odd, pattern OFF = even
    let required_parity: Vec<bool> = target_pattern.to_vec();
    
    // Find minimum parity presses needed (Part 1 style)
    let parity_toggles = required_parity.clone();
    let min_parity_presses = gaussian_elimination_gf2(&button_matrix, &parity_toggles);
    
    if let Some(min_presses) = min_parity_presses {
        // Helper to generate combinations
        fn generate_combinations(n: usize, k: usize) -> Vec<Vec<usize>> {
            if k == 0 {
                return vec![vec![]];
            }
            if k > n {
                return vec![];
            }
            if k == n {
                return vec![(0..n).collect()];
            }
            let mut result = Vec::new();
            // Include first element
            for mut combo in generate_combinations(n - 1, k - 1) {
                combo.insert(0, 0);
                for i in 1..combo.len() {
                    combo[i] += 1;
                }
                result.push(combo);
            }
            // Exclude first element
            for mut combo in generate_combinations(n - 1, k) {
                for i in 0..combo.len() {
                    combo[i] += 1;
                }
                result.push(combo);
            }
            result
        }
        
        let mut best_result: Option<i32> = None;
        let max_parity_to_try = num_buttons;
        
        for k in min_presses as usize..=max_parity_to_try {
            // Pruning
            if let Some(best) = best_result {
                if k > best as usize {
                    break;
                }
            }
            
            for button_combo in generate_combinations(num_buttons, k) {
                // Simulate pressing these buttons once
                let mut resulting_joltages = joltages.to_vec();
                for &btn_idx in &button_combo {
                    for &light in &buttons[btn_idx] {
                        if light >= 0 && (light as usize) < num_lights {
                            resulting_joltages[light as usize] += 1;
                        }
                    }
                }
                
                // Check if parity matches
                let resulting_parity: Vec<bool> = resulting_joltages.iter().map(|&j| j % 2 == 1).collect();
                if resulting_parity == required_parity {
                    // Check if all remaining are even (can divide by 2)
                    if resulting_joltages.iter().all(|&j| j % 2 == 0) {
                        // Divide by 2 and recurse
                        let halved_joltages: Vec<i32> = resulting_joltages.iter().map(|&j| j / 2).collect();
                        if let Some(remaining_result) = solve_part2_ilp(buttons, &halved_joltages, target_pattern) {
                            let total = k as i32 + remaining_result * 2;
                            best_result = Some(best_result.map(|b| b.min(total)).unwrap_or(total));
                        }
                    }
                }
            }
        }
        
        if best_result.is_some() {
            return best_result;
        }
    }
    
    // Fallback: divide-by-2 optimization
    if joltages.iter().all(|&j| j % 2 == 0) && joltages.iter().any(|&j| j > 0) {
        let halved_joltages: Vec<i32> = joltages.iter().map(|&j| j / 2).collect();
        if let Some(result) = solve_part2_ilp(buttons, &halved_joltages, target_pattern) {
            return Some(result * 2);
        }
        return None;
    }
    
    // Final fallback: bounded DFS
    let max_joltage = joltages.iter().max().copied().unwrap_or(0);
    
    fn dfs(
        button_idx: usize,
        current_joltages: &[i32],
        presses_so_far: i32,
        best: Option<i32>,
        buttons: &[Vec<i32>],
        joltages: &[i32],
    ) -> Option<i32> {
        if button_idx >= buttons.len() {
            if current_joltages.iter().zip(joltages.iter()).all(|(a, b)| a == b) {
                return Some(best.map(|b| b.min(presses_so_far)).unwrap_or(presses_so_far));
            }
            return best;
        }
        
        if let Some(b) = best {
            if presses_so_far >= b {
                return best;
            }
        }
        
        let mut best_result = best;
        for presses in 0..=max_joltage {
            if let Some(b) = best {
                if presses_so_far + presses >= b {
                    break;
                }
            }
            
            let mut new_joltages = current_joltages.to_vec();
            for &light in &buttons[button_idx] {
                if light >= 0 && (light as usize) < joltages.len() {
                    new_joltages[light as usize] += presses;
                }
            }
            
            if new_joltages.iter().zip(joltages.iter()).any(|(a, b)| a > b) {
                continue;
            }
            
            let result = dfs(button_idx + 1, &new_joltages, presses_so_far + presses, best_result, buttons, joltages);
            if let Some(r) = result {
                best_result = Some(best_result.map(|b| b.min(r)).unwrap_or(r));
            }
        }
        
        best_result
    }
    
    let initial_joltages = vec![0; num_lights];
    dfs(0, &initial_joltages, 0, None, buttons, joltages)
}

fn solve(input_data: &str) -> (String, String) {
    let lines: Vec<&str> = input_data.trim().lines().collect();
    if lines.is_empty() {
        return ("0".to_string(), "0".to_string());
    }
    
    let mut part1_total = 0;
    let mut part2_total = 0;
    
    for line in lines {
        if line.trim().is_empty() {
            continue;
        }
        
        let parsed = parse_line(line);
        if parsed.target_pattern.is_empty() {
            continue;
        }
        
        let num_lights = parsed.target_pattern.len();
        
        // Part 1: GF(2) linear system
        // Build incidence matrix: matrix[i][j] = true if button i toggles light j
        let mut button_matrix: Vec<Vec<bool>> = vec![vec![false; num_lights]; parsed.buttons.len()];
        for (i, btn) in parsed.buttons.iter().enumerate() {
            for &light in btn {
                if light >= 0 && (light as usize) < num_lights {
                    button_matrix[i][light as usize] = true;
                }
            }
        }
        
        // Target: all start OFF, need to toggle to match pattern
        // Required toggles: target_pattern (1 = needs toggle, 0 = no toggle)
        let required_toggles = parsed.target_pattern.clone();
        
        if let Some(result) = gaussian_elimination_gf2(&button_matrix, &required_toggles) {
            part1_total += result;
        }
        
        // Part 2: Integer Linear Programming
        if parsed.joltages.len() == num_lights {
            if let Some(result2) = solve_part2_ilp(&parsed.buttons, &parsed.joltages, &parsed.target_pattern) {
                part2_total += result2;
            }
        }
    }
    
    (part1_total.to_string(), part2_total.to_string())
}

fn main() {
    let data = fs::read_to_string("../data/input.txt").expect("Error reading file");
    let (part1, part2) = solve(&data);
    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
